DigitalLearningSolutions.Data/Services/GroupsService.cs

```diff
@@ -0,0 +1,234 @@
+            DelegateUser oldDelegateDetails,
+            AccountDetailsData newDelegateDetails,
+            CentreAnswersData newCentreAnswers,
+            string baseUrl
+        )
+        {
+            var changedLinkedFields = LinkedFieldHelper.GetLinkedFieldNumbersWithValuesFromChangedAnswers(
+                oldDelegateDetails,
+                newCentreAnswers,
+                jobGroupsDataService
+            );
+
+            var allSynchronisedGroupsAtCentre =
+                GetSynchronisedGroupsForCentre(oldDelegateDetails.CentreId).ToList();
+
+            foreach (var changedAnswer in changedLinkedFields)
+            {
+                var groupToRemoveDelegateFrom = allSynchronisedGroupsAtCentre.SingleOrDefault(
+                    g => g.LinkedToField == changedAnswer.LinkedFieldNumber && g.GroupLabel == changedAnswer.OldValue
+                );
+
+                var groupToAddDelegateTo = allSynchronisedGroupsAtCentre.SingleOrDefault(
+                    g => g.LinkedToField == changedAnswer.LinkedFieldNumber && g.GroupLabel == changedAnswer.NewValue
+                );
+
+                using var transaction = new TransactionScope();
+                if (groupToRemoveDelegateFrom != null)
+                {
+                    RemoveDelegateFromGroup(oldDelegateDetails.Id, groupToRemoveDelegateFrom.GroupId);
+                }
+
+                if (groupToAddDelegateTo != null)
+                {
+                    groupsDataService.AddDelegateToGroup(
+                        oldDelegateDetails.Id,
+                        groupToAddDelegateTo.GroupId,
+                        clockService.UtcNow,
+                        1
+                    );
+
+                    EnrolDelegateOnGroupCourses(
+                        oldDelegateDetails,
+                        newDelegateDetails,
+                        groupToAddDelegateTo.GroupId,
+                        baseUrl
+                    );
+                }
+
+                transaction.Complete();
+            }
+        }
+
+        public void EnrolDelegateOnGroupCourses(
+            DelegateUser oldDetails,
+            AccountDetailsData newDetails,
+            int groupId,
+            string baseUrl,
+            int? addedByAdminId = null
+        )
+        {
+            var groupCourses = groupsDataService.GetGroupCourses(groupId, oldDetails.CentreId);
+
+            foreach (var groupCourse in groupCourses)
+            {
+                DateTime? completeByDate = null;
+                if (groupCourse.CompleteWithinMonths != 0)
+                {
+                    completeByDate = clockService.UtcNow.AddMonths(groupCourse.CompleteWithinMonths);
+                }
+
+                var candidateProgressOnCourse =
+                    progressDataService.GetDelegateProgressForCourse(oldDetails.Id, groupCourse.CustomisationId);
+                var existingRecordToUpdate = candidateProgressOnCourse.FirstOrDefault(

```

Comment:
We should possibly be updating all we find, but I'll run it past Kevin tomorrow.

Reply:
From what I recall, the other records that would be found are completed/removed records that need to not be updated. We otherwise only have 1 record per delegate-course combination. Happy to be corrected though.

Reply:
Yep, checked with Kevin. He confirmed we should update all matching, in-progress Progress records rather than just the first.

Reply:
Well now I am very confused. Do we expect multiple in-progress Progress records per delegate-course combination?

Reply:
We don't, but the database doesn't constrain it. If we did end up with more than one, it's probably least likely to cause surprising behaviour if we just treat them all the same way.